python/advi.py

```diff
@@ -0,0 +1,89 @@
+import numba
+import numpy as np
+import pytensor
+import pytensor.tensor as pt
+
+
+def make_cost_func():
+    mu = pt.dvector("mu")
+    tau = pt.dvector("tau")
+    draws = pt.dmatrix("draws")
+    grads = pt.dmatrix("grads")
+    omega = pt.exp(-2 * tau)
+    expected_grads = -(draws - mu) * omega
+    cost = ((expected_grads - grads) ** 2 / omega).mean() / 2
+    d_mu, d_tau = pt.grad(cost, [mu, tau])
+    func = pytensor.function([mu, tau, draws, grads], [cost, d_mu, d_tau], mode="NUMBA")
+    return func
+
+
+cost_func = make_cost_func().vm.jit_fn
+
+
+@numba.njit(parallel=True)
+def compute_grads(draws):
+    out = np.empty_like(draws)
+    for i in numba.prange(len(draws)):
+        _, grad = logp_func(draws[i])
+        out[i] = grad
+    return out
+
+
+def advi(compiled_model, max_iter=10_000, residual=5e-3):
```

Comment:
don't we have ADVI at the pytensor level in pymc already?

Reply:
Sorry, this wasn't supposed to end up in the PR. I rebased